src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingAssessmentService.java

```diff
@@ -43,4 +57,85 @@ public Result saveManualAssessment(Result result, User assessor) {
+        newManualResult = saveManualAssessment(newManualResult, assessor);
+
+        if (submission.getParticipation() == null) {
+            newManualResult.setParticipation(submission.getParticipation());
+        }
+        Result savedResult = resultRepository.save(newManualResult);
+        savedResult.setSubmission(submission);
+
+        // Re-load result to fetch the test cases
+        newManualResult = resultRepository.findByIdWithEagerSubmissionAndFeedbackAndTestCasesElseThrow(newManualResult.getId());
+
+        if (submit) {
+            return submitManualAssessment(newManualResult, submission, participation, exercise);
+        }
+        return newManualResult;
+    }
+
+    private Result submitManualAssessment(Result newManualResult, ProgrammingSubmission submission, StudentParticipation participation, ProgrammingExercise exercise) {
+        newManualResult = resultRepository.submitManualAssessment(newManualResult);
+
+        if (participation.getStudent().isPresent()) {
+            singleUserNotificationService.checkNotificationForAssessmentExerciseSubmission(exercise, participation.getStudent().get(), newManualResult);
+        }
+
+        // Note: we always need to report the result over LTI, even if the assessment due date is not over yet.
+        // Otherwise it might never become visible in the external system
+        if (ltiNewResultService.isPresent()) {
+            ltiNewResultService.get().onNewResult((StudentParticipation) newManualResult.getParticipation());
+        }
+        if (ExerciseDateService.isAfterAssessmentDueDate(exercise)) {
+            resultWebsocketService.broadcastNewResult(newManualResult.getParticipation(), newManualResult);
+        }
+
+        sendFeedbackToAthena(exercise, submission, newManualResult.getFeedbacks());
+        handleResolvedFeedbackRequest(participation);
+        newManualResult.setParticipation(participation);
+
+        return newManualResult;
+    }
+
+    /**
+     * Send feedback to Athena (if enabled for both the Artemis instance and the exercise).
+     */
+    private void sendFeedbackToAthena(final ProgrammingExercise exercise, final ProgrammingSubmission programmingSubmission, final List<Feedback> feedbacks) {
+        if (athenaFeedbackSendingService.isPresent() && exercise.getFeedbackSuggestionsEnabled()) {
+            athenaFeedbackSendingService.get().sendFeedback(exercise, programmingSubmission, feedbacks);
+        }
+    }
+
+    private void handleResolvedFeedbackRequest(StudentParticipation participation) {
+        var exercise = participation.getExercise();
+        var isManualFeedbackRequest = exercise.getAllowManualFeedbackRequests() && participation.getIndividualDueDate() != null
+                && participation.getIndividualDueDate().isBefore(ZonedDateTime.now());
+        var isBeforeDueDate = exercise.getDueDate() != null && exercise.getDueDate().isAfter(ZonedDateTime.now());
```

Comment:
is it possible to use a helper function from `ExerciseDateService`? maybe `isBeforeEarliestDueDate`?

Reply:
No, we need to check the exercise due date here and explicitly want to ignore the individual due date of the participation. I  added a comment to explain that.